App: implement managing multiple projects (#107)

keggsmurph21 opened the pull request:
See issue #107 for more details

keggsmurph21 committed:
App: implement managing multiple projects (#107)

See the issue for more details

keggsmurph21 committed:
App: save `project_hash` after initializing a Project

keggsmurph21 commented on ultratrace2/model/files/bundle.py:
```diff
@@ -45,25 +45,27 @@ class FileBundleList:
         self.has_sound_impl = False
 
         self.current_bundle = None
+        self.bundles: List[FileBundle] = []  # FIXME: build up this data structure
```
Probably shouldn't be a `list`

calebho commented on ultratrace2/app.py:
```diff
@@ -15,10 +18,41 @@ def __init__(self, args):  # FIXME: be more granular here
 
-        self.project = Project(path)
+        # FIXME: allow changing the $HOME-as-root via command line arg?
+        self.ultratrace_home = os.path.join(os.environ["HOME"], ".ultratrace")
```
From what I can tell, Windows might not have this set. I think this does what you want on any platform
```suggestion
        self.ultratrace_home = os.path.expanduser(os.path.join("~", ".ultratrace"))
```

keggsmurph21 replied:
I'm unclear on whether we intend to support Windows, @jonorthwash is this a goal of ours?

keggsmurph21 replied:
In any case, 7c46ccb should supercede this

jonorthwash replied:
Yeah, in theory we'd like to support Windows.  But I still haven't fully tested on Windows (I did a little bit yesterday, but couldn't quite get it working—there was an issue with `libmagic` or something.)

jonorthwash replied:
Is the goal here to store app preferences in the user's home directory?

kmurphy4 replied:
So I think I'm abandoning storing anything in the home directory, since I don't think we'll need that. Maybe in the future it would be nice to store preferences (probably in `$XDG-*` if available), but I don't have a use-case yet.

Regarding `libmagic`, I think this is also obsolete as of [this commit](https://github.com/SwatPhonLab/UltraTrace/commit/04c8af2a13e61f0b80a46755ce1696a86288d81c), which removes the `python-magic` dependency in favor of the standard `mimetypes` lib.

calebho commented on ultratrace2/app.py:
```diff
@@ -15,10 +18,41 @@ def __init__(self, args):  # FIXME: be more granular here
+        # FIXME: allow changing the $HOME-as-root via command line arg?
+        self.ultratrace_home = os.path.join(os.environ["HOME"], ".ultratrace")
+        if not os.path.exists(self.ultratrace_home):
+            os.mkdir(self.ultratrace_home)
```
Should we set `mode=0o755` (or something else)? The default is `0o777`

keggsmurph21 replied:
Yeah, `0o755` is a more reasonable choice.  Didn't know `0o777` was the default ... I'll patch that in

keggsmurph21 replied:
Done in 1c15aca

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,40 @@
+
+    @classmethod
+    def initialize_from_path(cls, data_path) -> "Project":
+        project = Project(data_path)
```
In order to avoid half-initialized objects, add take an instance of `FileBundleList` in `__init__`, then in this alternate constructor, create a new instance of `FileBundleList`, call `scan_dir`, then pass it to the `Project` constructor.

Also, since this is a `classmethod`, you can use `cls` to invoke `__init__`:
```suggestion
        project = cls(data_path)
```

keggsmurph21 replied:
Specifically in reference to this comment, do you mind taking a look at https://github.com/SwatPhonLab/UltraTrace/commit/7c46ccb94bf866a69bef140d71a2d4a29de64b24 ?

calebho commented on ultratrace2/model/files/bundle.py:
```diff
@@ -45,25 +45,27 @@ class FileBundleList:
     )
 
-    def __init__(self, path: str, extra_exclude_dirs: List[str] = []):
+    def __init__(self, extra_exclude_dirs: Sequence[str] = []):
```
I think this is okay in this case, but in general you really shouldn't have mutable objects as defaults (c.f. https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/)
```suggestion
    def __init__(self, extra_exclude_dirs: Optional[Sequence[str]] = None):
```

keggsmurph21 replied:
(superceded)

calebho reviewed (commented):
Tests? :)

keggsmurph21 committed:
Project: change constructor pattern & save location

Following feedback on #107, this patch implements saving the ultratrace
metadata directly in the directory containing the source files.  In
addition, it changes the way we construct Projects and FileBundleLists
so that we avoid half-constructed objects.

keggsmurph21 committed:
Create directories in 0o755 mode (instead of default 0o777)

keggsmurph21 committed:
Project: remove old comments

keggsmurph21 committed:
Project: check if root_path is a directory

calebho commented on ultratrace2/model/files/bundle.py:
```diff
@@ -41,30 +41,43 @@ class FileBundleList:
+        self.has_sound_impl: bool = False
 
-        self.current_bundle = None
+        for filename, bundle in bundles.items():
```
```suggestion
        for bundle in bundles.values():
```

calebho commented on ultratrace2/model/files/bundle.py:
```diff
@@ -41,30 +41,43 @@ class FileBundleList:
+    ) -> "FileBundleList":
+
+        # FIXME: implement `extra_exclude_dirs` as a command-line arg
+        exclude_dirs = cls.exclude_dirs.copy().union(extra_exclude_dirs)
```
`.union()` returns a new set (c.f. https://docs.python.org/3.7/library/stdtypes.html#frozenset.union)
```suggestion
        exclude_dirs = cls.exclude_dirs.union(extra_exclude_dirs)
```

keggsmurph21 replied:
The doc you linked is for `FrozenSet`

keggsmurph21 replied:
Although it probably would be better to use `FrozenSet` anyway ...

keggsmurph21 replied:
Done f0f6154

calebho commented on ultratrace2/model/files/bundle.py:
```diff
@@ -41,30 +41,43 @@ class FileBundleList:
-        for path, dirs, filenames in os.walk(path, topdown=True):
-            dirs[:] = [d for d in dirs if d not in self.exclude_dirs]
+        for path, dirs, filenames in os.walk(root_path, topdown=True):
+            dirs[:] = [d for d in dirs if d not in exclude_dirs]
```
Correct me if I'm wrong, but I don't think `dirs` is used after this

keggsmurph21 replied:
Check the comment directly above this loop -- we're actually _modifying_ `dirs` here, not creating a copy.  This allows us to prune entire branches from the directory tree as we go.

calebho commented on ultratrace2/model/files/bundle.py:
```diff
@@ -83,12 +96,4 @@ def __init__(self, path: str, extra_exclude_dirs: List[str] = []):
-                self.has_image_impl = True
-            if not self.has_sound_impl and bundle.sound_file.has_impl():
-                self.has_sound_impl = True
+        return FileBundleList(bundles)
```
```suggestion
        return cls(bundles)
```

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
+                return project
+        except Exception as e:
+            logger.error(e)
+            raise RuntimeError(str(e))
```
Any reason to force everything to become a `RuntimeError`? Why not just
```suggestion
            raise
```

keggsmurph21 replied:
I want to specifically catch one type of error in the caller.  I figured `RuntimeError` was the most generic `Exception` subclass.

calebho replied:
In that case you can still do `except Exception` at the caller without erasing the type of the originally thrown exception

keggsmurph21 replied:
Done in 7450611

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
+            raise RuntimeError(str(e))
+
+    @classmethod
+    def get_by_path(cls, root_path) -> "Project":
```
```suggestion
    def get_by_path(cls, root_path: str) -> "Project":
```

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
     @classmethod
-    def load(cls):
-        raise NotImplementedError()
+    def load(cls, save_file) -> "Project":
```
```suggestion
    def load(cls, save_file: str) -> "Project":
```

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
+        root_path = os.path.realpath(os.path.abspath(root_path))  # absolute path
+        if not os.path.exists(root_path) or not os.path.isdir(root_path):
+            raise ValueError(
+                f"cannot initialize project at {root_path}: not a directory"
```
Maybe want a different error message for each case?

keggsmurph21 replied:
Fair

keggsmurph21 replied:
2bee499

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
+                f"cannot initialize project at {root_path}: not a directory"
+            )
+
+        save_dir = Project.get_save_dir(root_path)
```
```suggestion
        save_dir = cls.get_save_dir(root_path)
```

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
+        if not os.path.exists(save_dir):
+            os.mkdir(save_dir, mode=0o755)
+
+        save_file = Project.get_save_file(root_path)
```
```suggestion
        save_file = cls.get_save_file(root_path)
```

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
+
+        save_file = Project.get_save_file(root_path)
+        try:
+            return Project.load(save_file)
```
```suggestion
            return cls.load(save_file)
```

calebho commented on ultratrace2/model/project.py:
```diff
@@ -1,23 +1,68 @@
+
+        traces = TraceList()
+        file_bundles = FileBundleList.build_from_dir(root_path)
+        return Project(traces, file_bundles)
```
```suggestion
        return cls(traces, file_bundles)
```

keggsmurph21 committed:
App: Add types to constructor args, allow "headless" mode

keggsmurph21 committed:
App: add tests

keggsmurph21 committed:
Trace+XHair: fix type errors when running `pytest`

keggsmurph21 committed:
Update ultratrace2/model/files/bundle.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
FileBundleList: Make `cls.exclude_dirs` a `FrozenSet`

keggsmurph21 committed:
Merge branch 'manage-multiple-projects' of https://github.com/swatphonlab/ultratrace into manage-multiple-projects

keggsmurph21 committed:
Update ultratrace2/model/files/bundle.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
Update ultratrace2/model/files/bundle.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
Update ultratrace2/model/project.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
Update ultratrace2/model/project.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
Project: Use more descriptive error messages in ::get_by_path()

keggsmurph21 committed:
Merge branch 'manage-multiple-projects' of https://github.com/swatphonlab/ultratrace into manage-multiple-projects

keggsmurph21 committed:
Update ultratrace2/model/project.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
Update ultratrace2/model/project.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
Update ultratrace2/model/project.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

calebho commented on ultratrace2/app.py:
```diff
@@ -6,27 +7,30 @@
 
 class App:
-    def __init__(self, args):  # FIXME: be more granular here
+    def __init__(self, args: Namespace):
```
I think accepting a `Namespace` here will lead to spaghetti in the future because there's no way to know what attributes exist on it. Right now it's fine because there are only two parameters we need, but as this number grows, this approach won't scale. I prefer accepting the actual parameters rather than an untyped parameter bag.
```suggestion
    def __init__(self, headless: bool = False, path: Optional[str] = None):
```

keggsmurph21 replied:
Yeah I was worried about this.  I think you're right that unpacking the `Namespace` immediately after creation and then passing the unpacked dict to this constructor is a better idea.  Good catch.

keggsmurph21 replied:
f7b7f7a

keggsmurph21 committed:
Update ultratrace2/model/project.py

Co-Authored-By: Caleb <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
App: be more explicit about allowed parameters to App constructor

keggsmurph21 committed:
App: change parameter name in one test

keggsmurph21 committed:
Project: add tests for ::load()

NB: These tests only test for INVALID input. There are currently no
    tests for VALID input (since that behavior is not defined yet).

keggsmurph21 committed:
nox: Also calculate test coverage when running pytest

keggsmurph21 committed:
FileBundle: Simplify logic of `has_*_impl` initialization

keggsmurph21 committed:
Files: simplify inheritance of file implementations

Still unresolved:
 - How can we specify the desired resolution order (and still allow it
   be flexible / extensible)?

keggsmurph21 committed:
Remove `python-magic` dependency in favor of `mimetypes`

Python already includes a standard library for working with mimetypes,
so it's unclear why we were ever using `python-magic`.

keggsmurph21 committed:
Registry: Implement plugin-based architecture for loading files

keggsmurph21 commented on ultratrace2/model/files/bundle.py:
```diff
@@ -13,32 +19,38 @@ class FileBundle:
-    def set_alignment_file(self, alignment_file: AlignmentFile) -> None:
+    def set_alignment_file(self, alignment_file: AlignmentFileLoader) -> None:
+        if self.alignment_file is not None:
+            logger.warning("Overwriting existing alignment file")
```
@calebho This is another reason why we might want some kind of resolution order. Consider the case where we're scanning a directory that contains `XYZ.wav` and `XYZ.mp3` (assuming we know how to handle both).  Clearly, we would prefer the losslessly encoded `wav` file, but under the current system he one we "end up with" depends only on the order of directory traversal.

calebho replied:
> Clearly, we would prefer the losslessly encoded wav file

Are you sure that this matches the user's preference? If all users always have this preference, why would they include the recording in both encodings?

I'm not sure how this relates to having more than one loader for a given extension. Perhaps I'm misunderstanding what you mean by "resolution order".

jonorthwash replied:
A user might have a lossy file around for easy playback / demonstration purposes.

For a real example of something similar, I usually use flac for things like this (lossless compression), but might have wavs around because that's what the original data was recorded in and I don't like keeping the original around.  (The choice between wav and flac would just be one of resources: speed (wav better?) or memory (flac better?).)

keggsmurph21 replied:
> I'm not sure how this relates to having more than one loader for a given extension. Perhaps I'm misunderstanding what you mean by "resolution order".

So by "resolution order", I mean two separate things:
 1) As Jonathan mentioned, `$user` has files `XYZ.flac` and `XYZ.wav`:
    - As we traverse the directory, we encounter both.
    - WLOG, let's say we encounter `XYZ.flac` first.
    - We retrieve `FLACFileLoader` from the registry and successfully load `XYZ.flac`.
    - We save it to `bundles["xyz"].sound_file`.
    - We continue traversing, and we encounter `XYZ.wav`.
    - We retrieve `WAVFileLoader` from the registry and succesfully load `XYZ.flac`.
    - We go to save it to `bundles["xyz"].sound_file`, but that's already been set.
    - Currently, we just print a `logger.warning` and overwrite it, but we might want to give certain loaders "preference" over others, so that we might always choose `FLAC`, for example.

 2) `$user` has the file `ABC.flac` and registers `MyFLACLoader` with mime types `["audio/x-flac", "audio/flac"]` and extensions `["flac"]`.
    - We traverse the directory and encounter `ABC.flac`.
    - We try to retrieve from the registry, but error out since it found two possible loaders, but, again, we might want to give certain loaders "preference" over others (presumably user-defined ones should be preferred, but then what happens if they try to register `MyOtherFLACLoader` with the same mime types / extensions?).

I think (1) is a real possibility, and (2) is also possible but less likely.  One solution I can see is to present the user with a dialog for choosing their "`FileBundle` associations" (i.e., which sound file corresponds to which alignment file corresponds to which image-set file) in the case of conflicts, but to resolve automatically where possible.

calebho replied:
> Currently, we just print a logger.warning and overwrite it, but we might want to give certain loaders "preference" over others, so that we might always choose FLAC, for example.

Ditto. For now I think it makes sense to hard code this preference, then expose it as a user configurable thing later.

> We try to retrieve from the registry, but error out since it found two possible loaders

Does it make sense to differentiate an `audio/x-flac` FLAC from an `audio/flac` FLAC? Aren't they the same?

> we might want to give certain loaders "preference" over others (presumably user-defined ones should be preferred, but then what happens if they try to register MyOtherFLACLoader with the same mime types / extensions?)

Does it make sense to have multiple loaders registered at once? We can have the notion of preference without keeping all possible preferences loaded at the same time.

keggsmurph21 replied:
> > Currently, we just print a logger.warning and overwrite it, but we might want to give certain loaders "preference" over others, so that we might always choose FLAC, for example.
> 
> Ditto. For now I think it makes sense to hard code this preference, then expose it as a user configurable thing later.

OK, at what level in the abstraction would be enforce that?  Give each `FileLoader` a `priority` class attribute and take the highest?

> > We try to retrieve from the registry, but error out since it found two possible loaders
> 
> Does it make sense to differentiate an `audio/x-flac` FLAC from an `audio/flac` FLAC? Aren't they the same?

Yes, they are equivalent.  But that's not the reason it found two ... each file only has one mime type.  So WLOG, `("audio/x-flac", "flac")` could map to `FLACLoader` (built-in) _and_ `MyFLACLoader` (user extension).

> > we might want to give certain loaders "preference" over others (presumably user-defined ones should be preferred, but then what happens if they try to register MyOtherFLACLoader with the same mime types / extensions?)
> 
> Does it make sense to have multiple loaders registered at once? We can have the notion of preference without keeping all possible preferences loaded at the same time.

I think probably it does not make sense.  We could emit a `logger.warning` if we're overwriting an existing `FileLoader` at register-time.

jonorthwash replied:
> Does it make sense to differentiate an `audio/x-flac` FLAC from an `audio/flac` FLAC? Aren't they the same?

Ftr, the reason we have both in there is because we were getting files identified as either.  I forget if it was platform-specific or file-specific, but we definitely needed it to respond to either.

calebho commented on ultratrace2/model/files/loaders/base.py:
```diff
@@ -0,0 +1,58 @@
+
+    @classmethod
+    @abstractmethod
+    def from_file(cls, path: str) -> "FileLoaderBase":
```
This won't typecheck properly for subclasses which return an instance of the subclass. You need to define a `TypeVar` above:
```python
Self = TypeVar("Self", bound="FileLoaderBase")
```
then constrain the return type using `cls`:
```suggestion
    def from_file(cls: Type[Self], path: str) -> Self:
```
see: https://github.com/python/mypy/issues/1212#issuecomment-219250132

keggsmurph21 replied:
> This won't typecheck properly for subclasses which return an instance of the subclass. You need to define a `TypeVar` above:
> 
> ```python
> Self = TypeVar("Self", bound="FileLoaderBase")
> ```
> 
> then constrain the return type using `cls`:

Ah, this is what I was trying to puzzle through yesterday -- I didn't realize that I had to add the type hint to the `from_file` signature.  Thanks.

> see: [python/mypy#1212 (comment)](https://github.com/python/mypy/issues/1212#issuecomment-219250132)

lmao, straight from GvR, nice.

keggsmurph21 replied:
Fixed in a452e98

calebho commented on ultratrace2/model/files/loaders/base.py:
```diff
@@ -0,0 +1,58 @@
+
+
+class FileLoaderBase(ABC):
+    def __new__(cls, path: str):
```
Do we really need to define `__new__`?

keggsmurph21 replied:
You think it would be better to catch the `FileLoadError` at the call site? That seems reasonable to me.

keggsmurph21 replied:
Done ~4f86115~ e6895ba

calebho commented on ultratrace2/model/files/loaders/base.py:
```diff
@@ -0,0 +1,58 @@
+            logger.error(e)
+            return None
+
+    def __init__(self, path: str):
```
Do we really want to share `__init__` for all subclasses? In `DICOMFileLoader` this interface is already violated because it takes `pixels` as well.

keggsmurph21 replied:
I just want to enforce that each subclass will have a `path: str` property.  Can I do that like this
```python
class FileLoaderBase(ABC):
    path: str
```
?

calebho replied:
Does an abstract property work?

keggsmurph21 replied:
Done 2669b54

calebho commented on ultratrace2/model/files/loaders/base.py:
```diff
@@ -0,0 +1,58 @@
+        number can then be used to "slice up" any accompanying Alignment or Sound
+        files.
+        """
+        pass
```
You don't need `pass` because triple-quoted strings are expressions.
```suggestion
```

calebho commented on ultratrace2/model/files/loaders/dicom.py:
```diff
@@ -0,0 +1,79 @@
+        try:
+            dicom = pydicom.read_file(path)
+        except pydicom.errors.InvalidDicomError as e:
+            raise FileLoadError(str(e))
```
You might want to use exception chaining so that you don't erase the stack trace from the caught exception (which may be useful for debugging), e.g.
```suggestion
            raise FileLoadError("my error message") from e
```
see the sixth paragraph of the introduction in https://docs.python.org/3/library/exceptions.html

keggsmurph21 replied:
Cool, didn't know about this feature. Done in 336f414

calebho commented on ultratrace2/model/files/registry.py:
```diff
@@ -0,0 +1,67 @@
+        loader_cls: A file loader which knows how to load files with the given file extensions and MIME types
+    """
+
+    global __extension_to_loaders_map
```
You don't need `global` if you're not doing assignment
```suggestion
```

keggsmurph21 replied:
Done in fa457a1

keggsmurph21 committed:
FileLoaderBase: constrain ::from_file return type

See https://github.com/SwatPhonLab/UltraTrace/pull/109#discussion_r360178018
for more context.

keggsmurph21 committed:
Update ultratrace2/model/files/loaders/base.py

Co-Authored-By: Caleb Ho <caleb.yh.ho@gmail.com>

keggsmurph21 committed:
DICOMLoader: raise FileLoadError from InvalidDicomError

keggsmurph21 committed:
Registry: remove unnecessary `global` statements

Apparently, `global` is only necessary for assignment, and these
functions only need to access the value.

keggsmurph21 committed:
Project: Don't funnel ::load() errors into RuntimeErrors

Instead, let the exceptions propagate up to the caller.  For one thing,
this allows us to write more specific tests :^)

keggsmurph21 committed:
FileBundle: Pass responsibility for handling FileLoadError to caller

keggsmurph21 committed:
FileLoaderBase: Require `path: str` abstract attribute

keggsmurph21 committed:
Merge branch 'master' into manage-multiple-projects
